### ISSUE TITLE
[clang] Define the `__cpp_consteval` macro now that support is complete

### ISSUE THREAD

**Issue by BertalanD** [llvm/llvm-project#57094]
Commit 3d2629dd3aab17098813c68b5b76bb864bc5e285 marked `consteval` support as completed for the Clang 15 release. However, the `__cpp_consteval` feature test macro is still not defined. If the feature is indeed production-ready, it would be nice to define this macro, both on `main` and the release branch.

https://github.com/llvm/llvm-project/blob/0b12f770241134865be39d73413466da8430f126/clang/lib/Frontend/InitPreprocessor.cpp#L679

cc @cor3ntin, @ChuanqiXu9

See also #53906

**Comment by cor3ntin** [comment]
There are still some bugs in the implementation so i think it made sense to be conservative in 15 but we should strive to add the macro in clang 16.

@AaronBallman

**Comment by AaronBallman** [comment]
> There are still some bugs in the implementation so i think it made sense to be conservative in 15 but we should strive to add the macro in clang 16.
> 
> @AaronBallman

I agree -- we're *close* to being complete, but I think there's still some outstanding bugs we should address before we provide the macro. That said, we might discover new bugs by enabling the macro and trying out some large projects (in case anyone is looking for an experiment to run and report back on).

**Comment by AaronBallman** [comment]
I did some investigating and we're not as close as I'd like to be. Specifically, there are a few outstanding cases from P1073R3 that we don't handle properly:
```
namespace P1073R3 {
consteval int f() { return 42; } // expected-note 3 {{declared here}}
consteval auto g() { return f; }
// FIXME: there should be no diagnostics associated with either h() or r.
consteval int h(int (*p)() = g()) { return p(); } // expected-error {{call to consteval function 'P1073R3::g' is not a constant expression}} \
                                                     expected-note {{declared here}} \
                                                     expected-note {{pointer to a consteval declaration is not a constant expression}}
constexpr int r = h();   // expected-note {{in the default initalizer of 'p'}}
constexpr auto e = g();  // expected-error {{call to consteval function 'P1073R3::g' is not a constant expression}} \
                            expected-error {{constexpr variable 'e' must be initialized by a constant expression}} \
                            expected-note 2 {{pointer to a consteval declaration is not a constant expression}}
} // namespace P1073R3

namespace P1073R3 {
struct N {
  constexpr N() {}
  N(N const&) = delete;
};

template<typename T> constexpr void bad_assert_copyable() { T t; T t2 = t; }
using ineffective = decltype(bad_assert_copyable<N>());

// bad_assert_copyable<N> is not needed for constant evaluation
// (and thus not instantiated)
template<typename T> consteval void assert_copyable() { T t; T t2 = t; }
using check = decltype(assert_copyable<N>());
// FIXME: this should give an error because assert_copyable<N> is instantiated
// (because it is needed for constant evaluation), but the attempt to copy t is
// ill-formed.
} // namespace P1073R3
```
There is some implementation divergence, but I believe Clang's behavior is the least conforming of the implementations tested: https://godbolt.org/z/soMs4aWGo

**Comment by AaronBallman** [comment]
FWIW, I posted https://reviews.llvm.org/D144572 to try to update the public status page and get more info into the test suite as to what functionality is still missing before we should claim full support (IMO).

**Comment by Fznamznon** [comment]
> Specifically, there are a few outstanding cases from P1073R3 that we don't handle properly

I'm looking into these.

**Comment by Fznamznon** [comment]
Here is an attempt to fix the first one - https://reviews.llvm.org/D145251 .

**Comment by Fznamznon** [comment]
Regarding the following example:
```
namespace P1073R3 {
struct N {
  constexpr N() {}
  N(N const&) = delete;
};

template<typename T> constexpr void bad_assert_copyable() { T t; T t2 = t; }
using ineffective = decltype(bad_assert_copyable<N>());

// bad_assert_copyable<N> is not needed for constant evaluation
// (and thus not instantiated)
template<typename T> consteval void assert_copyable() { T t; T t2 = t; }
using check = decltype(assert_copyable<N>());
// FIXME: this should give an error because assert_copyable<N> is instantiated
// (because it is needed for constant evaluation), but the attempt to copy t is
// ill-formed.
} // namespace P1073R3
```

I think clang is correct here. https://wg21.link/p1937 proposes that in unevaluated contexts, consteval
functions should not be immediately evaluated. Clang implemented p1937 a while ago - https://github.com/llvm/llvm-project/commit/131b4620ee7847102479f399ce3e35a3c1cb5461 , so its behavior is correct here. I posted https://reviews.llvm.org/D145362 to update the test. It also adds examples from p1937.

@AaronBallman , is my understanding correct once the two outstanding cases (https://github.com/llvm/llvm-project/issues/57094#issuecomment-1440284125 ) are fixed, clang can claim support for consteval again?
I do realize there is some bugs still, but where is the place without bugs? :D

**Comment by Fznamznon** [comment]
I submitted https://reviews.llvm.org/D145251 and https://reviews.llvm.org/D145362 , looking if there is any remaining problems with examples from the standard.

**Comment by AaronBallman** [comment]
> I submitted https://reviews.llvm.org/D145251 and https://reviews.llvm.org/D145362 , looking if there is any remaining problems with examples from the standard.

I just checked the standard and didn't spot any other examples we weren't handling properly. I also looked through the open issues in Clang, and here are the ones that concern me still:

https://github.com/llvm/llvm-project/issues/60286 (missing copy elision during constant evaluation, rejects valid code)
https://github.com/llvm/llvm-project/issues/58207 (failed assertion due to invalid internal state)

There were a few others that could be worrisome, but they need further reduction and analysis to be sure. IMO, we don't have to be bug-free but we shouldn't knowingly reject or crash on valid code either. AFAIK, those are the only two outstanding issues that I think qualify as blockers for the feature test macro. If others know of issues they think should also block this, please speak up!

**Comment by Fznamznon** [comment]
> I also looked through the open issues in Clang, and here are the ones that concern me still:
 https://github.com/llvm/llvm-project/issues/60286 (missing copy elision during constant evaluation, rejects valid code)
 https://github.com/llvm/llvm-project/issues/58207 (failed assertion due to invalid internal state)

Ok, I'm looking into https://github.com/llvm/llvm-project/issues/58207 .

**Comment by h-vetinari** [comment]
> I submitted https://reviews.llvm.org/D145251 and https://reviews.llvm.org/D145362 , looking if there is any remaining problems with examples from the standard.

Notwithstanding the definition of the macro, it seems that after these PRs, the status of P1073R3 on https://clang.llvm.org/cxx_status.html should go back to "Full" (resp. "Unreleased")?

(I'm writing here as there's no separate issue for P1073 AFAICT)

**Comment by AaronBallman** [comment]
> > I submitted https://reviews.llvm.org/D145251 and https://reviews.llvm.org/D145362 , looking if there is any remaining problems with examples from the standard.
> 
> Notwithstanding the definition of the macro, it seems that after these PRs, the status of P1073R3 on https://clang.llvm.org/cxx_status.html should go back to "Full" (resp. "Unreleased")?
> 
> (I'm writing here as there's no separate issue for P1073 AFAICT)

Yes, in general I think the feature test macro and the feature status page should be in agreement. So once we define the macro, we should claim support on the webpage.

**Comment by h-vetinari** [comment]
My point was that (IIUC...) P1073 has a much smaller scope than all of C++20 `__cpp_consteval`; the latter may not be ready, the former should now be.

**Comment by h-vetinari** [comment]
> But if we think we've tracked those down already, maybe we can mark P1073 as full now.

TBH, I was naïvely going by the tests you had added in https://reviews.llvm.org/D144572, which have now been fixed. But that's possibly because I had the apparently wrong impression that the `__cpp_consteval` macro covers a superset of P1073.

**Comment by Fznamznon** [comment]
So, a little update
I've submitted fixes for
https://github.com/llvm/llvm-project/issues/61142
https://github.com/llvm/llvm-project/issues/58207 (and https://github.com/llvm/llvm-project/issues/59447 since this is technically a duplicate).
https://github.com/llvm/llvm-project/issues/60286

@AaronBallman , do you think we're ready to claim the full support again?

**Comment by AaronBallman** [comment]
> So, a little update I've submitted fixes for #61142 #58207 (and #59447 since this is technically a duplicate). #60286

Thank you for these fixes!

> @AaronBallman , do you think we're ready to claim the full support again?

Yes, I think we are. Would you mind putting up a patch to define `__cpp_consteval` and update the `www/cxx_status.html` page?

**Comment by Fznamznon** [comment]
> Yes, I think we are. Would you mind putting up a patch to define __cpp_consteval and update the www/cxx_status.html page?

Sure!
